[bug_fix] Updated README to match RedHat's ansible collection template (DCNE-198) (#692)
CiscoDevNet/ansible-aci: README.md

# ansible-aci

## Description

The ansible-aci project provides an Ansible collection for managing and automating your Cisco ACI environment. It consists of a set of modules and roles for performing tasks related to ACI.

> See the [Cisco.Aci collection index](https://galaxy.ansible.com/ui/repo/published/cisco/aci/content/) for a full list of modules and plugins.

Reviewer: ```suggestion
See the [cisco.aci collection index](https://galaxy.ansible.com/ui/repo/published/cisco/aci/content/) for a full list of modules and plugins.
```

Author: Done

## Requirements

- Ansible v2.15 or newer
- Python v3.10 or newer

> *Note: This collection is not compatible with versions of Ansible before v2.8.*

Reviewer: I think the requirement above of v2.15 and newer is probably overriding this note. Should we remove it?

Author: True, I thought it was a bit unnecessary.

Author: Removed

Follow the [Installing Ansible](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html) guide for detailed instructions.

## Installation

Before using this collection, you need to install it with the Ansible Galaxy command-line tool:

```sh
ansible-galaxy collection install cisco.aci
```

> You can also include it in a requirements.yml file and install it with ansible-galaxy collection install -r requirements.yml, using the format:

Reviewer: put the cli into a quote block with sh type like above

Author: Done

```yaml
collections:
  - name: cisco.aci
```

Note that if you install any collections from Ansible Galaxy, they will not be upgraded automatically when you upgrade the Ansible package.
To upgrade the collection to the latest available version, run the following command:

```sh
ansible-galaxy collection install cisco.aci --upgrade
```

> You can also install a specific version of the collection, for example, if you need to downgrade when something is broken in the latest version (please report an issue in this repository). Use the following syntax to install version 1.0.0:

Reviewer: The `for example, if you need to downgrade when something is broken in the latest version` does not sound great. I would rephrase that.

Author: This is taken right from the RedHat example but I'll change it. We don't need to follow it exactly.

Author: Re-worded. I took out the specific example of why someone would potentially attempt to downgrade.

```sh
ansible-galaxy collection install cisco.aci:==1.0.0
```

See [using Ansible collections](https://docs.ansible.com/ansible/devel/user_guide/collections_using.html) for more details.

### Latest Build

> Follow these instructions to get the latest collection build.

Reviewer: ```suggestion
Follow these instructions to get the latest collection by building it from the source code.
```

# ansible-aci

## Description

The ansible-aci project provides an Ansible collection for managing and automating your Cisco ACI environment. It consists of a set of modules and roles for performing tasks related to ACI.

This collection supports ACI 4.2+

See the [Cisco.Aci collection index](https://galaxy.ansible.com/ui/repo/published/cisco/aci/content/) for a full list of modules and plugins.

## Requirements

Ansible v2.15 or newer

*Note: This collection is not compatible with versions of Ansible before v2.8.*

## Installation

> Before using this collection, you need to install it with the Ansible Galaxy command-line tool:

Author: This installation part was part of the RedHat template. I can add some further links and mention Python 3.10+ as a requirement.

> Before using this collection, you need to install it with the Ansible Galaxy command-line tool:

Author: The parent page of `Installing collections` is already down below. But I am adding the Ansible instrallation guide here too.

> Before using this collection, you need to install it with the Ansible Galaxy command-line tool:

Author: Added the ansible install guide above.

# ansible-aci

The ansible-aci project provides an Ansible collection for managing and automating your Cisco ACI environment. It consists of a set of modules and roles for performing tasks related to ACI.
## Description

This collection has been tested and supports ACI 3.2+.
Modules supporting new features introduced in ACI API in specific ACI versions might not be supported in earlier ACI releases.
The `ansible-aci` project provides an Ansible collection for managing and automating your Cisco ACI environment. It consists of a set of modules and roles for performing tasks related to ACI.

> *Note: This collection is not compatible with versions of Ansible before v2.8.*

Reviewer 2: I think, we should still keep this note somewhere in readMe

Author: We already mention in the requirements section that the collection requires Ansible v2.15+
v2.8 has been EOL for almost 3 years now. I think its okay to remove.

# ansible-aci

## Description

> The `ansible-aci` project provides an Ansible collection for managing and automating your Cisco ACI environment. It consists of a set of modules and roles for performing tasks related to ACI.

Reviewer 3: I think the general norm is to expand an acronym the first time it is mentioned in a document after which only mentioning the acronym will do.

The `ansible-aci` project provides an Ansible collection for managing and automating your Cisco Application Centric Infrastructure (ACI) environment. It consists of a set of modules and roles for performing tasks related to ACI.

Author: Done 👍

```

### Second Approach
## Use Cases

Once the collection is installed, you can use it in a playbook by specifying the full namespace path to the module, plugin and/or role.

### Adding a new EPG

Go to [ansible-aci Actions](https://github.com/CiscoDevNet/ansible-aci/actions)
```yml
- hosts: aci
  gather_facts: no

Select the latest CI build
  tasks:
  - name: Add a new EPG
    cisco.aci.aci_epg:
      hostname: apic
      username: admin
      password: SomeSecretPassword
      tenant: production
      ap: intranet
      epg: web_epg
      description: Web Intranet EPG
      bd: prod_bd
    delegate_to: localhost
```

Under Artifacts download collection and unzip it using Terminal or Console.
## Optimizing Playbooks

*Note: The collection file is a zip file containing a tar.gz file. We recommend using CLI because some GUI-based unarchiver might unarchive both nested archives in one go.*
There are two main methods to optimize the execution of ACI modules in your playbooks.

Install the unarchived tar.gz file
1. Using the ACI HTTPAPI plugin
1. Using the `suppress_` options

```sh
ansible-galaxy collection install cisco-aci-1.0.0.tar.gz —-force

> ```

Reviewer: Why removing this second approach?

Author: @akinross and I agreed, that since the builds aren't consistently green due to the integration tests that we wouldn't recommend doing it this way anymore.
https://github.com/CiscoDevNet/ansible-aci/pull/692#discussion_r1799281736
The idea was to add an nightly dev build that gets published instead. Not sure if this is possible with Ansible Galaxy but thought I would add a backlog item for it.

Reviewer: How would the nightly be different from the CI actions downloads and how is it different from building from source in regard to the integration tests?

How are we tracking this nightly suggestion?

Until we have the nightly available more easily I would keep that in.

Author: Its a good point, there would not be much difference. The nightly build would just be pushing the latest master which gets built in the CI every day anyway.

I have added the second approach back in.

> - 4.2(7s)
> - 5.2(7g)
> - 6.0(2h)

Reviewer: I think we need to be a bit more vague on versions as I don't want to update this every time we change the setups.
```suggestion
- 4.2
- 5.2
- 6.0
```

## Contributing

Ongoing development efforts and contributions to this collection are tracked as issues in this repository.

We welcome community contributions to this collection. If you find problems, need an enhancement or need a new module, please open an issue or create a PR against the [Cisco ACI collection repository](https://github.com/CiscoDevNet/ansible-aci/issues).

## Support

This collection supports any ACI version within the Last Day of Support (LDOS) date.

Some modules and options within the collection are only available from specific versions of ACI. The versions a module or option supports are documented within the individual module documentation.

To find EOL announcements for ACI versions, refer to the [End-of-Life and End-of-Sale Notices](https://www.cisco.com/c/en/us/products/cloud-systems-management/application-policy-infrastructure-controller-apic/eos-eol-notice-listing.html) page.

## Release Notes

See the [Changelog](CHANGELOG.rst) for full release notes.

## Related Information

> For additional information and guides see the [Cisco ACI DevNet documentation](https://developer.cisco.com/docs/aci/ansible/#cisco-aci-ansible-modules).

Reviewer: Lets also add a link to : 
* https://developer.cisco.com/docs/nexus-as-code/aci-with-ansible/
* https://developer.cisco.com/learning/tracks/aci-programmability/ansible-aci-intro/

ansible-galaxy collection install cisco-aci-* --force
```

#### Second Approach

> Go to [ansible-aci Actions](https://github.com/CiscoDevNet/ansible-aci/actions)

Reviewer 4: Should we remove this part from the docs, since actions have not been running stable.

Reviewer 4: Perhaps we can push a nightly build on the ansible galaxy page: https://galaxy.ansible.com/ui/repo/published/cisco/aci/

Author: I agree with removing this part. The builds do produce a collection to download but due to the current integration tests they are always red.
Not sure anyone would be comfortable using those.
Pushing a nightly build can be a later task.

Author: Removed this part

ansible-galaxy collection install cisco-aci-1.0.0.tar.gz —-force
```

## Use Cases

Once the collection is installed, you can use it in a playbook by specifying the full namespace path to the module, plugin and/or role.

### Adding a new EPG

```yml
- hosts: aci
  gather_facts: no

  tasks:
  - name: Add a new EPG
    cisco.aci.aci_epg:
      hostname: apic
      username: admin
      password: SomeSecretPassword
      tenant: production
      ap: intranet
      epg: web_epg
      description: Web Intranet EPG
      bd: prod_bd

>     delegate_to: localhost

Reviewer 4: Do we want to delegate to localhost in this example, don't think we require this anymore with httpapi plugin. Should we make a mention of this plugin in the readme?

Author: It is mentioned in Optimizing Playbooks but I can add a short summary like you mentioned in the other comment.

Author: I will keep the delegate option here because this is what we list in all the module examples currently.

```

## Optimizing Playbooks

> To find out more about optimizing playbook execution, please refer to the [Optimizing Playbooks](docs/optimizing.md) documentation.

Reviewer 4: should we provide short summary of the contents found here?

Author: Can do 👍

Author: Done

## Testing

Integration tests for each module in the cisco.aci collections are executed against the following ACI versions:

- 4.2(7s)
- 5.2(7g)
- 6.0(2h)

## Contributing

Ongoing development efforts and contributions to this collection are tracked as issues in this repository.

We welcome community contributions to this collection. If you find problems, need an enhancement or need a new module, please open an issue or create a PR against the [Cisco ACI collection repository](https://github.com/CiscoDevNet/ansible-aci/issues).

## Support

> This collection has supports the ACI versions listed in the Testing section.

Author: Hmm LDOS for 3.2 is 30th Nov 2024. Did we remove testing this version too early? We can't really say we support any ACI version that is within LDOS. However, I'll change it anyway because its only in a month or so.

> This collection has supports the ACI versions listed in the Testing section.

Author: Re-worded this section and removed the explicit mention of 4.2+ support.

> Modules supporting new features introduced in ACI API in specific ACI versions might not be supported in earlier ACI releases.

Reviewer 4: This sentence reads a bit weird for me. Would mention something like: Arguments exposed in the cisco.aci collection modules are only supported when ACI supports the feature.

Author: Yeah, this is a sentence we had in the readme before. 🤔 
I can re-word it.

Author: Updated
